packages/components/menu/src/Menu.tsx

```diff
@@ -0,0 +1,226 @@
+        availableElements.length > 0 &&
+        focusedIndex < availableElements.length
+      ) {
+        (availableElements[focusedIndex] as HTMLElement).focus({
+          preventScroll: false,
+        });
+      }
+    }
+  }, [isOpen, focusedIndex]);
+
+  const closeAndFocusTrigger = useCallback(() => {
+    handleClose();
+    triggerRef.current?.focus({ preventScroll: true });
+  }, [handleClose]);
+
+  const handleMenuListKeyDown = useCallback(
+    (event: React.KeyboardEvent) => {
+      if (event.key === 'Tab') {
+        event.preventDefault();
+        closeAndFocusTrigger();
+        return;
+      }
+
+      handleArrowsKeyDown(event);
+    },
+    [closeAndFocusTrigger, handleArrowsKeyDown],
+  );
+
+  const contextValue: MenuContextType = useMemo(
+    () => ({
+      getTriggerProps: (_props = {}, _ref = null) => ({
+        ..._props,
+        onClick: (event) => {
+          if (isOpen) {
+            handleClose();
+          } else {
+            handleOpen();
+          }
+          _props.onClick?.(event);
+        },
+        ref: mergeRefs(triggerRef, _ref),
+      }),
+      getMenuListProps: (_props = {}, _ref = null) => ({
+        ..._props,
+        ref: mergeRefs(menuListRef, _ref),
+        onKeyDown: (event) => {
+          handleMenuListKeyDown(event);
+          _props.onKeyDown?.(event);
+        },
+      }),
+      getMenuItemProps: (_props = {}) => ({
+        ..._props,
+        onClick: (event) => {
+          _props.onClick?.(event);
+          if (closeOnSelect) {
+            handleClose();
+          }
+        },
+      }),
+    }),
+    [isOpen, handleMenuListKeyDown, closeOnSelect, handleClose, handleOpen],
+  );
+
+  return (
+    <MenuContextProvider value={contextValue}>
+      <Popover
+        {...props}
+        isOpen={isOpen}
+        onClose={closeAndFocusTrigger}
+        // eslint-disable-next-line jsx-a11y/no-autofocus
+        autoFocus={false}
+      >
+        {props.children}
+      </Popover>
+    </MenuContextProvider>
+  );
+}
+
+type UseMenuOpenStateProps = Pick<
+  MenuProps,
+  'isOpen' | 'defaultIsOpen' | 'onOpen' | 'onClose'
+>;
+
+const useMenuOpenState = (props: UseMenuOpenStateProps) => {
+  const { isOpen, defaultIsOpen, onOpen, onClose } = props;
+  const [isOpenState, setIsOpen] = useState(defaultIsOpen || false);
+
+  const isControlled = isOpen !== undefined;
+  const isOpenValue = isControlled ? isOpen : isOpenState;
+
+  const handleClose = useCallback(() => {
+    if (!isControlled) {
+      setIsOpen(false);
+    }
+    onClose?.();
+  }, [isControlled, onClose]);
+
+  const handleOpen = useCallback(() => {
+    if (!isControlled) {
+      setIsOpen(true);
+    }
+    onOpen?.();
+  }, [isControlled, onOpen]);
+
+  return { isOpen: isOpenValue, isControlled, handleClose, handleOpen };
+};
+
+interface UseArrowKeyNavigationProps {
+  itemsContainerRef: React.MutableRefObject<HTMLElement>;
+  itemsSelector: string;
+}
+
+const useArrowKeyNavigation = ({
```

Comment:
Worth extracting? https://github.com/contentful/forma-36/discussions/1008

Reply:
I guess, we'll extract this if we'll be using it in any other component.